Make irb a railties dependency

st0012 opened the pull request:
### Motivation / Background

This PR proposes adding irb to the dependencies of railties for the following reasons:

- Firstly, including irb as a dependency allows users with older versions of Ruby to benefit from the latest version of irb, instead of being limited to the version bundled with their Ruby installation.

- Additionally, there is an [ongoing discussion](https://bugs.ruby-lang.org/issues/19351) to move irb to bundled gems. If this change is implemented, users may need to declare irb in their Gemfile in order to use it for Rails console. By adding irb as a dependency, users can avoid the extra effort of manually specifying irb in their Gemfile.


(**Old Description**)
<details><summary>Details</summary>
<p>

This PR updates app generator's template to add `gem "irb"` to new Rails applications' Gemfile. This is for several reasons:

1. irb constantly receives improvements and bug fixes, so differences between 2 Ruby versions' irb could be significant. For example, Ruby 3.2's irb (v1.6.2), has 10+ new commands than Ruby 3.1's irb (v1.4.1). 
    By installing irb as a gem, Ruby 3.1's Rails users can get the same features Ruby 3.2 users get when they run `rails console`.
1. Having irb in Gemfile means `dependabot` users can get irb updated automatically.
1. It makes users aware of that they can control the version of irb they use.

(BTW, there's a (still under discussion) [plan](https://bugs.ruby-lang.org/issues/19351) to move `irb` to bundled gems.)

</p>
</details> 

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

yahonda commented:
Thank you for opening a pull request.

While I understand that irb is developed rapidly, I am unsure if adding irb to the gemfile of a Rails application is a good idea.

As far as I understand, `irb` is one of the default gems that ships with Ruby and is updated as newer versions of Ruby are released. 

If users want to update the default gems, they can add them to their application's gemfile. Users can choose to use the latest one or the one that ships with Ruby.

I'm a little worried that this pull request will cause "Let's add most of the default gems to Rails Gemfile," which will make Rails applications Gemfile busy.

skipkayhil commented:
Maybe a good compromise would be to add it as a dependency of railties? I think that would cover all of the points raised above, and I think it would be more accurate than adding it to the Gemfile because it is really a dependency of `rails console` (and not applications):

https://github.com/rails/rails/blob/1fd0bc1094144b268a3b931a92eb6aeb32fe5c7b/railties/lib/rails/commands/console/console_command.rb#L3

yahonda commented:
> Maybe a good compromise would be to add it as a dependency of railties? 

I am afraid I do not think so because `irb` is one of the `default` gems. It means users do not need to add any dependency in Gemfile or gemspec. It differs from `bundled` gems that requires adding dependency in Gemfile or gemspec.

st0012 commented:
> If users want to update the default gems, they can add them to their application's gemfile. Users can choose to use the latest one or the one that ships with Ruby.

It's also true to `debug`. But Rails already has it in the Gemfile to make sure:
- Users don't need to upgrade to `3.1` to use it
- And, because its under rapid development, it helps keeping it up-to-date

> I'm a little worried that this pull request will cause "Let's add most of the default gems to Rails Gemfile," which will make Rails applications Gemfile busy.

I'm not adding it because it's a default gem. I'm adding it because `irb` is a tool users directly interact with through Rails console, in all environments. Unlike most default gems, any bug fixes/features to irb could be felt by users right away.

Another important point is that users can still use Ruby 2.7 or 3.0 to create new Rails apps. And their irb versions will lag even further behind, and there's no  benefit to stay at those versions.

I think my point is: if you agree that having up-to-date irb is important and beneficial to Rails users, shouldn't Rails help them keep it up-to-date by simply having it in the Gemfile? I think this fits the "convention over configuration" ideology.

st0012 commented:
> Maybe a good compromise would be to add it as a dependency of railties?

There are Pros and Cons on this compared to the Gemfile approach:

Pros:
- It means Rails can control the irb version users use, which could make Rails easier to integrate with irb.
   For example, if irb implemented [command extension APIs](https://github.com/ruby/irb/issues/513), it'd simplify Rails-side's implementation if that version of irb is required.
- It doesn't add another entry to users' Gemfile.

Cons:
- AFAIK, dependabot wouldn't bump gems not listed in Gemfile.

In the long-term, it should be a better approach. I didn't propose it first because:

- There isn't a feature that'd benefit Rails from this yet. I want to implement command extensions in irb and update/implement Rails commands through it this year. But it's not happening yet.
- It feels more intrusive than updating Gemfile's template.

But if you think it's ok, we can go with it too.

> I am afraid I do not think so because irb is one of the default gems. It means users do not need to add any dependency in Gemfile or gemspec. It differs from bundled gems that requires adding dependency in Gemfile or gemspec.

FYI, there's a [plan](https://bugs.ruby-lang.org/issues/19351) to move `irb` to bundled gems.

ghiculescu commented on railties/lib/rails/generators/rails/app/templates/Gemfile.tt:
```diff
@@ -32,6 +32,10 @@ gem "bootsnap", require: false
 # gem "rack-cors"
 <%- end -%>
+
+# Installing irb as a gem helps keeping it up-to-date
```
My gut feel is the same as yahonda's here. But if we do go ahead with this, I think we should think of a different comment here. This explains to other Rails developers why irb was added, but I don't think it says much to Rails users. Compare to most of the other comments in the Gemfile template, which describe why the gem is there and what you can use it for.


```suggestion
# irb powers the Rails console [https://guides.rubyonrails.org/command_line.html#bin-rails-console]
```

st0012 replied:
That's a good call, thank you 👍

hahmed commented:
> I'm adding it because irb is a tool users directly interact with through Rails console, in all environments.

You can configure the console to use something [other than IRB](https://github.com/rails/rails/blob/main/guides/source/configuring.md#configconsole) example below, not sure for that reason IRB should be added to the app gemfile. 

```
console do
  # this block is called only when running console,
  # so we can safely require pry here
  require "pry"
  config.console = Pry
end
```

st0012 commented:
@hahmed Sorry I don't get the argument. Follow the same logic, users can chose `unicorn` as the web server too, so `puma` shouldn't be at Gemfile either? 🤔 

Adding `irb` to Gemfile simply means "it's a great default to use, and if you don't want it you can remove it".

hahmed commented:
If we take database as an example, if you pass in --database=postgresql to the rails new generator, mysql is not added to the gemfile. In this case, if you bundle IRB with the app gemfile, then the user has to remove IRB from the gemfile if they want to use pry.

The argument about webservers is different because from what I can see, you cannot change (I thought you could) the default or recommended webserver (puma) via the rails new generator, therefore it's added to the app gemfile. If IRB is the default and recommended choice, like puma then yes this change makes sense.

ghiculescu commented:
@hahmed you can't set `pry` as the console through the generator, can you? So I don't think the metaphor really applies here. The user also has to add `pry` to the gemfile, so they can remove `irb` at the same time.

matthewd commented:
> FWIW, there's a [plan](https://bugs.ruby-lang.org/issues/19351) to move irb to bundled gems.

We should address that list, but I think irb should be a railties dependency. Every line we add to the Gemfile is [generated] configuration over convention, and should be viewed with skepticism. I don't think there's enough benefit to allowing people to omit it, to justify allowing the application to find itself without a way to start a console.

rafaelfranca commented:
I'm with Matthew on this one. I think it should be a railties dependency.

yahonda commented:
I'm also fine adding irb as railties dependency.

tenderlove commented:
Making it a Railties dependency makes sense.

st0012 committed:
Add irb to railties' dependencies

- Firstly, including irb as a dependency allows users with older versions
  of Ruby to benefit from the latest version of irb, instead of being
  limited to the version bundled with their Ruby installation.

- Additionally, there is an [ongoing discussion](https://bugs.ruby-lang.org/issues/19351) to move irb to bundled gems.
  If this change is implemented, users may need to declare irb in their
  Gemfile in order to use it for Rails console. By adding irb as a dependency,
  users can avoid the extra effort of manually specifying irb in their Gemfile.

st0012 commented on railties/railties.gemspec:
```diff
@@ -44,6 +44,7 @@ Gem::Specification.new do |s|
   s.add_dependency "rake", ">= 12.2"
   s.add_dependency "thor", "~> 1.0"
   s.add_dependency "zeitwerk", "~> 2.6"
+  s.add_dependency "irb", "~> 1.6"
```
I'm not sure how strict/loose this should be, so I selected the latest minor release. But I don't see a reason to let users specify anything older.

st0012 replied:
More thoughts about this:

- If we set the upper bound, we'd restrict users' ability to upgrade `irb` in the given Rails version. Ideally, users should be able to upgrade to `2.0` if they feel like it.
- If we don't set the upper bound, users may unintentionally upgrade to a `irb` that has breaking change.

I think we could:

- Declare it as a dependency with: `>= 1.6`
- And add it to the Gemfile template with `gem "irb", "~> 1.6"`

WDYT?

rafaelfranca replied:
If we use any feature that only exists in irb 1.6 we should have this. Otherwise, just allow any version to be used.

st0012 replied:
Make sense. Updated 👍

st0012 commented:
Thanks for all your feedback 🙏 
It looks like the current majority supports making it a railties' dependency, so I'll update this PR.

hahmed commented:
> @hahmed you can't set pry as the console through the generator, can you? So I don't think the metaphor really applies here. The user also has to add pry to the gemfile, so they can remove irb at the same time.

Sure you can, firstly create the template file (https://guides.rubyonrails.org/rails_application_templates.html) then run that through the generator (`rails new example -m template.rb --api`), here is an example https://gist.github.com/hahmed/d0e7155f17cc700dab99e44328f649f8

zzak commented on railties/railties.gemspec:
```diff
@@ -44,6 +44,7 @@ Gem::Specification.new do |s|
   s.add_dependency "rake", ">= 12.2"
   s.add_dependency "thor", "~> 1.0"
   s.add_dependency "zeitwerk", "~> 2.6"
+  s.add_dependency "irb"
```
@st0012 Is there a reasonable base version we want to prefer? Like `>= 1.0.0"` or similar? I know for sure rubygems will warn about these, which is annoying but ignorable -- however we may want to ensure user doesn't accidentally use an old broken version somehow. If that's possible given irb is a default gem? 🤔

yahonda replied:
FYI https://github.com/rails/rails/pull/47442#discussion_r1119398688

Edgarruiz85 commented:
Duplicate of # #47594

tierra commented:
I'm only just now getting to upgrades that include this change, on a project using Sorbet.

Given IRB is still included in the version of Ruby I'm using (3.2.2), the `IRB::ReadlineInputMethod` class is derived from `InputMethod`. However, in IRB 1.8.0+, the `IRB::ReadlineInputMethod` class is now derived from `StdioInputMethod`, as ruby/irb#671.

This is tripping up Sorbet with both versions being loaded:

```
sorbet/rbi/gems/irb@1.8.3.rbi:2456: Parent of class IRB::ReadlineInputMethod redefined from IRB::InputMethod to IRB::StdioInputMethod https://srb.help/5012
    2456 |class IRB::ReadlineInputMethod < ::IRB::StdioInputMethod
                                           ^^^^^^^^^^^^^^^^^^^^^^^

sorbet/rbi/gems/irb@1.8.3.rbi:2532: Parent of class IRB::RelineInputMethod redefined from IRB::InputMethod to IRB::StdioInputMethod https://srb.help/5012
    2532 |class IRB::RelineInputMethod < ::IRB::StdioInputMethod
                                         ^^^^^^^^^^^^^^^^^^^^^^^
```

My only solution to this for now is explicitly pinning back the `irb` gem to `1.7.4`, but at least this does work (as long as Railties doesn't require a newer version).

It seems odd for Railties to introduce this as a required dependency to me, rather than leaving it up to applications to optionally do so for themselves, especially since Railties isn't doing this as a requirement for a newer version of IRB for some specific feature.

st0012 commented:
This is an issue on Sorbet's packaging of default gems' rbi files, not an issue on either IRB or Rails. You can resolve it by excluding IRB in Tapioca's config as [explained here](https://github.com/ruby/irb/issues/723#issuecomment-1762353753).

> Railties isn't doing this as a requirement for a newer version of IRB for some specific feature.

IRB is the base of Rails console and in many cases IRB features/bugs are also Rails console's, at least from users' perspective. If we look at it this way, declaring it as a dependency will help users install the latest IRB and get the best experience, instead of getting stuck with older versions with known issues & lack of features. For example, it will make all Rails 7.2 users get #47413's upstream fix now, instead of waiting until they upgrade to Ruby 3.3.

jarthod commented:
Hi, I just noticed this change while upgrading to Rails 7.1 because it broke my CI and deployment due to the many other dependencies brought in by `irb`. Mainly: `rdoc  → psych → stringio`.

1. Being forced to have `psych` in my gemfile now means I have to install `libyaml-dev` on every machine otherwise the gem install fails.
2. And adding `stringio` causes [confusion between version when using phusion passenger](https://stackoverflow.com/questions/77322464/rails-7-1-1-spawning-error-occurred-passenger-you-have-already-activated-s).

Of course we could all spend some days trying to fix and workaround these situations, but I'm wondering if the addition of `irb` as a dependency is really that worth it to cause this many impacts on setup?
